File: Java/BOJ/practice/2792 보석상자/main.py
# 학생은 항상 같은 생삭의 보석만 가져간다
N, M = map(int, input().split())
gems = []
left, right = 1,0

for i in range(M):
    gems.append(int(input()))
    right = max(right, gems[i])
print(gems)

answer = right

while (left <= right):
    mid = (left + right) // 2
    totalStudent = 0

    for gem in gems:
        totalStudent += (gem // mid)
        if gem % mid:
            totalStudent += 1
    
    if totalStudent <= N:
        answer = mid
        right = mid - 1
    else:
        left = mid + 1

print(answer)

File: Java/BOJ/practice/2792 보석상자/Main.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int M = Integer.parseInt(input[1]);

        List<Integer> gems = new ArrayList<>();
        int left = 1;
        int right = 0;

        for (int i=0;i<M;i++) {
            int currentGem = Integer.parseInt(br.readLine());
            gems.add(currentGem);

            right = Math.max(right, currentGem);
        }
        int zilous = right;
        while (left <= right) {
            int mid = (left + right) / 2;
            int totalCount = 0;
            for (int i=0; i<M;i++) {
                totalCount += (gems.get(i) / mid);
                if (gems.get(i) % zilous != 0) totalCount += 1;
            }

            if (totalCount <= N) {
                zilous = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        System.out.println(zilous);
    }
}